packages/gatsby/src/commands/build-html.ts

```diff
@@ -83,18 +145,22 @@ const runWebpack = (
       // because of this line we can't use our watch helper
       // These things should use emitter
       compiler.hooks.invalid.tap(`ssr file invalidation`, () => {
-        needToRecompileSSRBundle = true
+        SSRBundleReceivedInvalidEvent = true
+        SSRBundleWillInvalidate = false // we were waiting for this event, we are no longer waiting for it
       })
 
+      let isFirstCompile = true
       const watcher = compiler.watch(
         {
           ignored: /node_modules/,
         },
         (err, stats) => {
           // this runs multiple times
-          needToRecompileSSRBundle = false
```

Comment:
setting it to `false` here was possibly resetting any `invalid` events that happen during compilation, so this was moved to place where recompilation is triggered